### PR TITLE
apptainer: unbreak --nv (userns+nvliblist.conf variant)

### LINKED ISSUE #230851: singularity cannot use nvidia gpu.
### Describe the bug
singularity cannot use nvidia gpu because it does not access 'nvidia-container-cli'

### Steps To Reproduce
1. add `programs.singularity.enable=true`
2. run `singularity exec --nvccli docker://nvidia/cuda:10.0-base nvidia-smi`

### Expected behavior
return result of `nvidia-smi`

### Screenshots
image

### Additional context
This is because the path is not listed in singularity.conf, as also in a.
It is caused by neither the option `--nv` nor `--nvccli`.
I tried to handle it with override, but since it was called in `callPackage`, I think it is impossible to handle.
Also, I created a conf file in `/usr/local/etc/singularity/` as shown in the singularity docs, but it was not read.

### Notify maintainers
@jbedo  @ShamrockLee 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.3.0, NixOS, 23.05 (Stoat), 23.05.20230430.da45bf6`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

### PR DESCRIPTION
**Pull request by SomeoneSerge**
## Description of changes

Supersedes (the apptainer part of) https://github.com/NixOS/nixpkgs/pull/279235 whose scope grew out of control

CC @misumisumi @ShamrockLee I'm going to merge as soon as Ofborg is green, unless there are any objections

Closes https://github.com/NixOS/nixpkgs/issues/230851

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/virtualization/singularity/generic.nix [hunk 1]
```diff
@@ -263,6 +280,37 @@ in
...
+      image-saxpy = callPackage
+        ({ singularity-tools, cudaPackages }:
+          singularity-tools.buildImage {
+            name = "saxpy";
+            contents = [ cudaPackages.saxpy ];
+            memSize = 2048;
+            diskSize = 2048;
+            singularity = finalAttrs.finalPackage;
+          })
+        { };
```
Comment (SomeoneSerge) [inline pkgs/applications/virtualization/singularity/generic.nix:292]: Forgetting now if it's an issue one the tests being unfree. Will wait for it to fail

**Comment by SomeoneSerge** [comment]
Ofborg didn't exactly "fail", it just silently skipped both image-saxpy and image-hello-cowsay: https://logs.ofborg.org/?key=nixos/nixpkgs.280076&attempt_id=61daa35f-9362-45c7-bdcd-71d8855ddda4

Moving the attribute out of tests, waiting for ofborg, and merging

File: pkgs/applications/virtualization/singularity/apptainer/0001-ldCache-patch-for-driverLink.patch [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ 	"path/filepath"
+ 	"regexp"
+ 	"strings"
+@@ -154,14 +155,49 @@ func Resolve(fileList []string) ([]string, []string, error) {
+ // lists three variants of libEGL.so.1 that are in different locations, we only
+ // report the first, highest priority, variant.
+ func ldCache() (map[string]string, error) {
++    driverDirs := strings.Split("@driverLink@/lib", ":")
++    if machine, err := elfMachine(); err == nil && machine == elf.EM_386 {
++        driverDirs = strings.Split("@driverLink@-32/lib", ":")
```
Comment (SomeoneSerge) [inline pkgs/applications/virtualization/singularity/apptainer/0001-ldCache-patch-for-driverLink.patch:28]: Afterthought: this is probably wrong in `pkgsi686Linux` (going to end up with `...-32-32`)

**Comment by trofi** [comment]
I think this change caused the eval failure as:

```
$ nix build --no-link -f. apptainer.gpuChecks.saxpy
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:9:12:
            8|
            9|   strict = derivationStrict drvAttrs;
             |            ^
           10|

       … while evaluating derivation 'run-apptainer-cuda-saxpy'
         whose name attribute is located at /home/slyfox/dev/git/nixpkgs-master/pkgs/stdenv/generic/make-derivation.nix:331:7

       … while evaluating attribute 'nativeBuildInputs' of derivation 'run-apptainer-cuda-saxpy'
         at /home/slyfox/dev/git/nixpkgs-master/pkgs/stdenv/generic/make-derivation.nix:375:7:
          374|       depsBuildBuild              = elemAt (elemAt dependencies 0) 0;
          375|       nativeBuildInputs           = elemAt (elemAt dependencies 0) 1;
             |       ^
          376|       depsBuildTarget             = elemAt (elemAt dependencies 0) 2;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: attribute 'image-saxpy' missing
       at /home/slyfox/dev/git/nixpkgs-master/pkgs/applications/virtualization/singularity/generic.nix:303:72:
          302|                 ''
          303|                   ${lib.getExe finalAttrs.finalPackage} exec --nv $@ ${finalAttrs.passthru.tests.image-saxpy} saxpy
             |                                                                        ^
          304|                 '';
```

**Comment by trofi** [comment]
Proposed the trivial fix as https://github.com/NixOS/nixpkgs/pull/296497

**Pull request merged**